test/cbmc/stubs/core_sntp_stubs.c

```diff
@@ -0,0 +1,185 @@
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file core_sntp_stubs.c
+ * @brief Definition of stubs for UDP transport and authentication interfaces of coreSNTP API.
+ */
+
+#include <stdint.h>
+#include "core_sntp_client.h"
+#include "core_sntp_serializer.h"
+#include "core_sntp_stubs.h"
+
+#define TEST_TIMESTAMP         \
+    {                          \
+        .seconds = UINT32_MAX, \
+        .fractions = 1000      \
+    }
+
+/* An exclusive bound on the times that the NetworkInterfaceSendStub will be
+ * invoked before returning a loop terminating value. This is usually defined
+ * in the Makefile of the harnessed function. */
+#ifndef MAX_NETWORK_SEND_TRIES
+    #define MAX_NETWORK_SEND_TRIES    2
+#endif
+
+/* An exclusive bound on the times that the NetworkInterfaceReceiveStub will
+ * return an unbound value. At this value and beyond, the
+ * NetworkInterfaceReceiveStub will return zero on every call. */
+#ifndef MAX_NETWORK_RECV_TRIES
+    #define MAX_NETWORK_RECV_TRIES    4
+#endif
+
+static SntpTimestamp_t testTime = TEST_TIMESTAMP;
+
+int32_t NetworkInterfaceReceiveStub( NetworkContext_t * pNetworkContext,
+                                     uint32_t serverAddr,
+                                     uint16_t serverPort,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    __CPROVER_assert( pBuffer != NULL,
+                      "NetworkInterfaceReceiveStub pBuffer is NULL." );
+
+    __CPROVER_assert( __CPROVER_w_ok( pBuffer, bytesToRecv ),
+                      "NetworkInterfaceReceiveStub pBuffer is not writable up to bytesToRecv." );
+
+    /* Clean the buffer so as to make it available to write data into it. */
+    __CPROVER_havoc_object( pBuffer );
+
+    int32_t bytesOrError;
+    static size_t tries = 0;
+
+    /* It is a bug for the application defined transport receive function to return
+     * more than bytesToRecv. */
+    __CPROVER_assume( bytesOrError <= ( int32_t ) bytesToRecv );
+
+    if( tries < ( MAX_NETWORK_RECV_TRIES - 1 ) )
+    {
+        tries++;
+    }
+    else
+    {
+        tries = 0;
+        bytesOrError = 0;
```

Comment:
I think bytesorerror is the number of bytes received by this try.  I think by setting it to 0 in this case, you are always assuming the final invocation is an error.  Is that what you want?  Maybe in the final invocation you want to allow the possibility that all bytes left were received? The same question comes to mind in other stubs.